app/ante/eth.go

```diff
@@ -117,25 +116,25 @@ func (avd EthAccountVerificationDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx
 
 		// check whether the sender address is EOA
 		fromAddr := common.BytesToAddress(from)
-		codeHash := avd.evmKeeper.GetCodeHash(fromAddr)
-		if codeHash != common.BytesToHash(evmtypes.EmptyCodeHash) {
-			return ctx, sdkerrors.Wrapf(sdkerrors.ErrInvalidType,
-				"the sender is not EOA: address <%v>, codeHash <%s>", fromAddr, codeHash)
+		acct, err := avd.evmKeeper.GetAccount(ctx, fromAddr)
+		if err != nil {
+			return ctx, sdkerrors.Wrapf(sdkerrors.Wrapf(sdkerrors.ErrInvalidType,
+				"the sender is not EthAccount: address <%v>", fromAddr), "")
 		}
-
-		acc := avd.ak.GetAccount(ctx, from)
-		if acc == nil {
-			acc = avd.ak.NewAccountWithAddress(ctx, from)
+		if acct == nil {
+			acc := avd.ak.NewAccountWithAddress(ctx, from)
 			avd.ak.SetAccount(ctx, acc)
+			acct = statedb.NewEmptyAccount()
+		} else if !bytes.Equal(acct.CodeHash, evmtypes.EmptyCodeHash) {
```

Comment:
use account Type

Reply:
the `acct` is actually not an `EthAccount` here, but `statedb.Account`, I added an `IsContract` method to it.